### PR TITLE
PAPER-66: Bandwidth Tuning paper reproduction

### PR DESCRIPTION
**Pull request by Tancredoss**
- Reproduction of the paper : On the similarity of bandwidth-tuned quantum kernels and classical kernels (https://arxiv.org/abs/2503.05602)
- No Readme.md for now. I'm currently writing it

Related Jira: PAPER-66

### CODE REVIEW

File: papers/Bandwidth_tunning/results/run_20260630-162953/done.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+Completed
```
Comment (ben9871) [inline papers/Bandwidth_tunning/results/run_20260630-162953/done.txt]: Remove this. Unnecessary file and shouldn't be generated during the runs

File: papers/Bandwidth_tunning/lib/metrics.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import merlin
+import torch
+
+
+def matrix_sqrt(A):
+    """Calcule la racine carrée matricielle d'une matrice symétrique semi-définie positive."""
+    # Décomposition en valeurs propres (A = V * L * V^T)
```
Comment (ben9871) [inline papers/Bandwidth_tunning/lib/metrics.py]: mmmmm. all of your comments are in french. Seeing as this repo is meant to be read by the wider community, you may have to switch these comments to english

File: papers/Bandwidth_tunning/lib/metrics.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        computation_space=merlin.ComputationSpace.FOCK,
+    )
+
+    if X_test is None:
+        return _fidelity_kernel(X_train)
+    else:
+        return _fidelity_kernel(X_test, X_train)
+
+def projected_fidelity_kernel(feature_map, X_train, X_test = None):
+    _projected_fidelity_kernel = merlin.ProjectedFidelityKernel(
```
Comment (ben9871) [inline papers/Bandwidth_tunning/lib/metrics.py]: your fork of reproduced papers and the running of this repo depends on having your specific branch of merlin which is not the case for most users. It is good that you have tried an implementation version and prototyped how it would look in MerLin. But you will have to put this code within the repo itself so that the code is self-contained and runnable. This repo should run with merlinquantum == 0.4.0

File: papers/Bandwidth_tunning/results/run_20260630-101339/run.log [hunk 4]
```diff
@@ -0,0 +1,3 @@
+2026-06-30 10:13:39 | INFO | runtime | Starting Bandwidth_tunning run
```
Comment (ben9871) [inline papers/Bandwidth_tunning/results/run_20260630-101339/run.log]: you have too many run-specific files. 
you have results and results contains folder with run_year/month/day. 
I would do outdir --> contains folders with the runs data everytime you run.
But then I would have a results directory which just contains the key figures generated from the latest run. you could have a 'generate_figures.py' to run after the results are obtained which then populates results. 
What should be on the github itself are key-findings, and how to reproduce them, not baggage.

File: data/Bandwidth_tunning/fashion_mnist/train-images-idx3-ubyte.gz
Comment (ben9871) [inline data/Bandwidth_tunning/fashion_mnist/train-images-idx3-ubyte.gz:1]: far too many data files added to the repo. This data is all very easily accessible via torchvision aswell https://docs.pytorch.org/vision/stable/datasets.html . you can gitignore the data. The default config could run with a very small dataset or something synthetic. This should be clearly documented too

File: papers/Bandwidth_tunning/run.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+# Reproduce all experiments from Johri et al. (2020)
+# "Nearest Centroid Classification on a Trapped Ion Quantum Computer"
```
Comment (ben9871) [inline papers/Bandwidth_tunning/run.sh]: you took a file from another repo and used it as basis which is fine but make sure the comments are aligned with the bandwidth tuning paper and the usage. Right now the documentation does not match the file

File: papers/Bandwidth_tunning/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# On the similarity of bandwidth-tuned quantum kernels and classical kernels
+# Paper : Roberto Flórez-Ablan, Marco Roth, and Jan Schnabel (2025) arXiv:2503.05602
+
+scikit-learn>=1.9.0
+matplotlib>=3.8.0
+git+https://github.com/Tancredoss/merlin
```
Comment (ben9871) [inline papers/Bandwidth_tunning/requirements.txt]: it must work from an existing merlin installation on pip. merlinquantum==0.4.0 . It will have to be changed to match this

File: papers/Bandwidth_tunning/lib/runner.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        # Size of the training and testing datasets
+        NB_TRAIN = exp["train_sample"]
+        NB_TEST = exp["test_sample"]
+
+        SEEDS = np.random.default_rng(seed).integers(low=0, high=10000, size=exp["nb_seeds"])
+
+        print(f"experiment {exp['description']} running")
+        X_train, y_train, X_test, y_test = data(cfg["dataset"]["name"])
+
+        for seed in range(exp["nb_seeds"]):
```
Comment (ben9871) [inline papers/Bandwidth_tunning/lib/runner.py:211]: you are doing PCA on X_train, y_train, X_test and y_test here.
and you put these in a loop.
but you rebind them with the same names, so if you did dim 80, the next seed would just be doing PCA on size 80 which won't reduce again. you must check this and make sure the seed is being properly applied. seems not to be

File: papers/Bandwidth_tunning/lib/metrics.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    # 4. Norme spectrale (la plus grande valeur singulière, équivalente à ord=2)
+    spectral_norm = torch.linalg.matrix_norm(inner_matrix, ord=2)
+    # 5. Racine carrée finale de la formule
+    g = torch.sqrt(spectral_norm)
+    return g
+
+
+def calculate_eta_max(K):
+    L, V = torch.linalg.eigh(K)
+    eta_max = L[-1]
```
Comment (ben9871) [inline papers/Bandwidth_tunning/lib/metrics.py]: should this be normalised?

**Review by ben9871 — CHANGES_REQUESTED** [review]
Aside from the remarks discussed in the comments here, there is some files missing and formatting issues. https://github.com/merlinquantum/reproduced_papers/blob/main/how_to_contribute.md . The notebook, README.md and tests are missing from the repo which makes it fail smoke tests and reduces how informative it is for a new user. That page should be read again to make sure the missing things are added

File: papers/Bandwidth_tunning/results/run_20260630-162953/done.txt (hunk 1 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/results/run_20260630-162953/done.txt]: done

File: papers/Bandwidth_tunning/lib/metrics.py (hunk 2 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/lib/metrics.py]: done in theory

File: papers/Bandwidth_tunning/results/run_20260630-101339/run.log (hunk 4 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/results/run_20260630-101339/run.log]: done

File: data/Bandwidth_tunning/fashion_mnist/train-images-idx3-ubyte.gz
Comment (Tancredoss) [reply data/Bandwidth_tunning/fashion_mnist/train-images-idx3-ubyte.gz:1]: done

File: papers/Bandwidth_tunning/run.sh (hunk 5 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/run.sh]: done

File: papers/Bandwidth_tunning/requirements.txt (hunk 6 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/requirements.txt]: done

File: papers/Bandwidth_tunning/lib/runner.py (hunk 7 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/lib/runner.py:211]: done

File: papers/Bandwidth_tunning/lib/metrics.py (hunk 8 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/lib/metrics.py]: I don't think so, i'll check the paper

Comment (Tancredoss) [reply papers/Bandwidth_tunning/lib/metrics.py]: after review it had to be normalised

Comment (Tancredoss) [reply papers/Bandwidth_tunning/lib/metrics.py]: done

File: papers/Bandwidth_tunning/lib/metrics.py (hunk 3 above)
Comment (Tancredoss) [reply papers/Bandwidth_tunning/lib/metrics.py]: done
